blog/2020-11-10-streaming-vitess-at-bolt.adoc

```diff
@@ -0,0 +1,181 @@
+====
+++++
+<img src="/images/vitess/vitess_connector_multi_shards.png" style="max-width:100%;" class="responsive-image">
+++++
+*Figure 3. Each connector subscribes to all shards of a specific keyspace*
+====
+
+*Option 2:*
+
+As shown in _Figure 4_, each connector instance captures change events from a specific keyspace/shard pair. The connector instance gets the initial (the current) VGTID  position of the keyspace/shard pair from VTCtld gRPC, which is another Vitess component. Each connector instance, independently, uses the VGTID it gets to subscribe to VStream gRPC and continuously capture change events from VStream and sends them to Kafka. To support the Vitess Reshard operation, you would need more manual operations.
+
+[.centered-image.responsive-image]
+====
+++++
+<img src="/images/vitess/vitess_connector_single_shard.png" style="max-width:100%;" class="responsive-image">
+++++
+*Figure 4. Each connector subscribes to one shard of a specific keyspace*
+====
+
+Internally, each connector task uses a gRPC thread to constantly receive change events from VStream and puts the events into an internal blocking queue. The connector task thread polls events out of the queue and sends them to Kafka, as can be seen in _Figure 5_.
+
+[.centered-image.responsive-image]
+====
+++++
+<img src="/images/vitess/vitess_connector_internal.png" style="max-width:100%;" class="responsive-image">
+++++
+*Figure 5. How each connector task works internally*
+====
+
+=== Replication Challenges
+
+While we’re implementing the Vitess Connector and digging deeper into Vitess, we’ve also realized a few challenges.
+
+==== Vitess Reshard
+
+The Vitess connector supports the Vitess Reshard operation when the connector is configured to subscribe to all shards of a given keyspace. VStream sends a VGTID that contains the shard GTID for all shards. Vitess Resharding is transparent to users. Once it’s completed, Vitess will send the VGTID of the new shards. Therefore, the connector will use the new VGTID after reshard. However, you need to make sure that the connector is up and running when the reshard operation takes place. Especially please check that the offset topic of the connector has the new VGTID before deleting the old shards. This is because in case the old shards are deleted, VStream will not be able to recognize the VGTID from the old shards.
+
+If you decide to subscribe to one shard per connector, the connector does not provide out-of-the-box support for Vitess resharding. One manual workaround to support resharding is creating one new connector per target shard. For example, one new connector for the `commerce/-80` shard, and another new connector for the `commerce/80-` shard. Bear in mind that because they’re new connectors, by default, new topics will be created, however, you can use link:https://debezium.io/documentation/reference/configuration/topic-routing.html[Debezium logical topic router] to route the records to the same kafka topics.
```

Comment:
```suggestion
If you decide to subscribe to one shard per connector, the connector does not provide out-of-the-box support for Vitess resharding. One manual workaround to support resharding is creating one new connector per target shard. For example, one new connector for the `commerce/-80` shard, and another new connector for the `commerce/80-` shard. Bear in mind that because they’re new connectors, by default, new topics will be created, however, you could use the link:https://debezium.io/documentation/reference/configuration/topic-routing.html[Debezium logical topic router] to route the records to the same Kafka topics.
```